### ISSUE TITLE
Surprising concretization choices with invalid variants in external packages

### ISSUE THREAD

**Issue by Jordan474** [LLNL/spack#39036]
### Steps to reproduce

I think defining non-default-value variants in specs of external packages has quite surprising side effects on the concretization.

Here is an example of `vtk+qt` chosing **`qt@3.3.8b` instead of `qt@5.15.10`** when you configure `elfutils ~debuginfod` as external.

Starting with a couple configurations files:

```console
$ cat elfutils-external/packages.yaml 
packages:
  elfutils:
    buildable: False
    externals:
    - spec: elfutils@0.176
      prefix: /usr
$ cat elfutils-external-nodebuginfod/packages.yaml 
packages:
  elfutils:
    buildable: False
    externals:
    - spec: elfutils@0.176 ~debuginfod
      prefix: /usr
```

I get:

```console
$ spack spec --fresh vtk+qt ^elfutils~debuginfod | grep qt@
 -       ^qt@5.15.10%gcc@12.1.0~dbus~debug~doc~examples~framework~gtk+gui+opengl~phonon+shared+sql+ssl+tools~webkit build_system=generic patches=2081e9c,51aeba5,75bcb42,7f34d48,84b0991,8cd4be9,9378afd arch=linux-ubuntu22.04-skylake
$ spack -C elfutils-external spec --fresh vtk+qt | grep qt@
 -       ^qt@5.15.10%gcc@12.1.0~dbus~debug~doc~examples~framework~gtk+gui+opengl~phonon+shared+sql+ssl+tools~webkit build_system=generic patches=2081e9c,51aeba5,75bcb42,7f34d48,84b0991,8cd4be9,9378afd arch=linux-ubuntu22.04-skylake
$ spack -C elfutils-external-nodebuginfod spec --fresh vtk+qt | grep qt@
 -       ^qt@3.3.8b%gcc@12.1.0~dbus~debug~doc~examples~framework~gtk+gui+opengl~phonon+shared+sql+ssl+tools~webkit build_system=generic patches=92b0b28,a00f475 arch=linux-ubuntu22.04-skylake
```

The behavior also happens with spack 0.20, but **not in spack 0.19**.

Notes:
- I need external elfutils for external tooling compatibility
- I don't have debuginfod installed, so some package fail to install if they expect elfutils+debuginfod
- It took me quite a while before narrowing down to this example. The actual "bug" in my software stack was not so quite trivial to understand.

### Information on your system

* **Spack:** 0.21.0.dev0 (a99eaa954150778624f23459bafd49ce0397b099)
* **Python:** 3.10.12
* **Platform:** linux-ubuntu22.04-skylake
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by Jordan474** [comment]
This is even more surprising that without external it's ok:

```console
$ spack spec --fresh vtk+qt ^elfutils~debuginfod | grep qt@
 -       ^qt@5.15.10%gcc@12.1.0~dbus~debug~doc~examples~framework~gtk+gui+opengl~phonon+shared+sql+ssl+tools~webkit build_system=generic patches=2081e9c,51aeba5,75bcb42,7f34d48,84b0991,8cd4be9,9378afd arch=linux-ubuntu22.04-skylake
```

**Comment by Jordan474** [comment]
**Ok, I think this is my mistake:**

elfutils's  debuginfod variant is only available when `@0.179:`. So when I specify **an impossible `elfutils@0.176 ~debuginfod`**, then Spack tries very hard to not use it, and it either switches to an old qt to avoid glib, or it switches to libelf...

(I should know, it's not the first time I got surprised by a conditional variant, last time was matplotlib+image...)

I'm still seeing other unwanted side effects due to my externals configuration, but I think I'm going to go and check all my external specs validity first...
